yatsm/pipeline/_topology.py

```diff
@@ -111,11 +114,27 @@ def _config_to_deps(config, dsk=None, overwrite=True):
                 logger.debug('Task already computed and not overwrite - not '
                              'adding: {}'.format(task))
 
-        # If this task didn't provide any new data/record, cull it
+        # TODO: maybe we don't want to cull tasks, because they
+        #           * might have important side effects (e.g., IO)
+        #           * overwrite data (e.g., masking provides and
+        #             requires `data: ['nir']`)
         if not task_needed:
-            logger.debug('Culling task {} because everything it provides is '
-                         'already calculated (e.g., from cache)'.format(task))
-            del dsk[task]
+            if not prov:
```

Comment:
Tasks like "write this data to disk" aren't "pure" in a sense because of the file-writing side effect that can't be represented as a "provided" result in the `Pipe`. To accommodate, I've put this check in so we don't cull tasks that don't claim they output anything.